src/LOAuth.ts

```diff
@@ -111,16 +123,39 @@ class LOAuth {
     state = JSON.stringify(state);
 
     /**
-     * Skip over the encoding when there is a 'state' query parameter.  This is the
-     * scenario when we are loading from the redirected auth endpoint and the state
-     * is already encoded in the URL parameter.  The problem is that encoding/decoding
-     * is being handled on multiple layers so that this client loads after the auth
-     * redirect, client-oauth will parse the state from the query param as unencoded,
-     * stringified JSON and when we pass an encoded version in, there is a comparison
-     * failure which results in a second redirect.
+     * When the auth flow is initiated there will be no 'state' query parameter.
+     * Because it is stringified JSON it needs to be encoded before passing it to the
+     * auth provider. It ends up being doubly encoded. The redirect URI will look like:
+     * https://*.amazoncognito.com/oauth2/authorize?client_id=CLIENTID&state=%257B%2522projectId%2522%253A%2522test%2522%257D
+     * If we don't encode it here, there will only be a single level of encoding, which
+     * will break handoff to a second auth provider (cognito -> google):
+     * https://*.amazoncognito.com/oauth2/authorize?identity_provider=Google&redirect_uri=https://localhost:8080&response_type=CODE&client_id=CLIENTID&state={"projectId":"test"}&scope=openid
+     * The above link results in a 400 because of the state query parameter.
+     *
+     * If the user flows through a third party auth sign in, they will be redirected to:
+     * https://localhost:8080/?code=CODE&state=%7B%22projectId%22%3A%22test%22%7D
+     * Notice only a single encoding this time.
+     *
+     * If the user is already authenticated and we are redirected immediately from an
+     * auth server (no login page seen), the redirect contains the state value but it
+     * is doubly encoded:
+     * https://localhost:8080/?code=CODE&state=%257B%2522projectId%2522%253A%2522test%2522%257D
+     *
+     * We need to return a matching 'state' value or the auth flow will fail and the
+     * user will be redirected again.  When comparing the state value, reading the
+     * query string removes one layer of encoding so we either need to return an
+     * unencoded string (for the case of a singly encoded query parameter) on a singly
+     * encoded string (for the case of no query parameter or a doubly encoded one)
```

Comment:
I feel like this entire block could be removed, and would be better served as a PR comment.  It seems like it's describing whatever encoding issue we worked through last.

Reply:
I agree generally, but I wanted this to be here since on its own the code doesn't really make that much sense.  But if we have the base64 encoding none of this matters!